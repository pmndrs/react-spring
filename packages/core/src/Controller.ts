import { OneOrMore, UnknownProps, Lookup, Falsy } from '@react-spring/types'
import {
  is,
  raf,
  each,
  noop,
  flush,
  toArray,
  eachProp,
  flushCalls,
  addFluidObserver,
  FluidObserver,
} from '@react-spring/shared'

import { getDefaultProp } from './helpers'
import { FrameValue } from './FrameValue'
import { SpringRef } from './SpringRef'
import { SpringValue, createLoopUpdate, createUpdate } from './SpringValue'
import { getCancelledResult, getCombinedResult } from './AnimationResult'
import { runAsync, RunAsyncState, stopAsync } from './runAsync'
import { scheduleProps } from './scheduleProps'
import {
  AnimationResult,
  AsyncResult,
  ControllerFlushFn,
  ControllerUpdate,
  OnChange,
  OnRest,
  OnStart,
  SpringValues,
} from './types'

/** Events batched by the `Controller` class */
const BATCHED_EVENTS = ['onStart', 'onChange', 'onRest'] as const

let nextId = 1

/** Queue of pending updates for a `Controller` instance. */
export interface ControllerQueue<State extends Lookup = Lookup>
  extends Array<
    ControllerUpdate<State, any> & {
      /** The keys affected by this update. When null, all keys are affected. */
      keys: string[] | null
    }
  > {}

export class Controller<State extends Lookup = Lookup> {
  readonly id = nextId++

  /** The animated values */
  springs: SpringValues<State> = {} as any

  /** The queue of props passed to the `update` method. */
  queue: ControllerQueue<State> = []

  /**
   * The injected ref. When defined, render-based updates are pushed
   * onto the `queue` instead of being auto-started.
   */
  ref?: SpringRef<State>

  /** Custom handler for flushing update queues */
  protected _flush?: ControllerFlushFn<this>

  /** These props are used by all future spring values */
  protected _initialProps?: Lookup

  /** The counter for tracking `scheduleProps` calls */
  protected _lastAsyncId = 0

  /** The values currently being animated */
  protected _active = new Set<FrameValue>()

  /** The values that changed recently */
  protected _changed = new Set<FrameValue>()

  /** Equals false when `onStart` listeners can be called */
  protected _started = false

  private _item?: any

  /** State used by the `runAsync` function */
  protected _state: RunAsyncState<this> = {
    paused: false,
    pauseQueue: new Set(),
    resumeQueue: new Set(),
    timeouts: new Set(),
  }

  /** The event queues that are flushed once per frame maximum */
  protected _events = {
    onStart: new Map<
      OnStart<SpringValue<State>, Controller<State>, any>,
      AnimationResult
    >(),
    onChange: new Map<
      OnChange<SpringValue<State>, Controller<State>, any>,
      AnimationResult
    >(),
    onRest: new Map<
      OnRest<SpringValue<State>, Controller<State>, any>,
      AnimationResult
    >(),
  }

  constructor(
    props?: ControllerUpdate<State> | null,
    flush?: ControllerFlushFn<any>
  ) {
    this._onFrame = this._onFrame.bind(this)
    if (flush) {
      this._flush = flush
    }
    if (props) {
      this.start({ default: true, ...props })
    }
  }

  /**
   * Equals `true` when no spring values are in the frameloop, and
   * no async animation is currently active.
   */
  get idle() {
    return (
      !this._state.asyncTo &&
      Object.values(this.springs as Lookup<SpringValue>).every(
        spring => spring.idle
      )
    )
  }

  get item() {
    return this._item
  }

  set item(item) {
    this._item = item
  }

  /** Get the current values of our springs */
  get(): State & UnknownProps {
    const values: any = {}
    this.each((spring, key) => (values[key] = spring.get()))
    return values
  }

  /** Set the current values without animating. */
  set(values: Partial<State>) {
    for (const key in values) {
      const value = values[key]
      if (!is.und(value)) {
        this.springs[key].set(value)
      }
    }
  }

  /** Push an update onto the queue of each value. */
  update(props: ControllerUpdate<State> | Falsy) {
    if (props) {
      this.queue.push(createUpdate(props))
    }
    return this
  }

  /**
   * Start the queued animations for every spring, and resolve the returned
   * promise once all queued animations have finished or been cancelled.
   *
   * When you pass a queue (instead of nothing), that queue is used instead of
   * the queued animations added with the `update` method, which are left alone.
   */
  start(props?: OneOrMore<ControllerUpdate<State>> | null): AsyncResult<this> {
    let { queue } = this as any
    if (props) {
      queue = toArray<any>(props).map(createUpdate)
    } else {
      this.queue = []
    }

    if (this._flush) {
      return this._flush(this, queue)
    }

    prepareKeys(this, queue)
    return flushUpdateQueue(this, queue)
  }

  /** Stop all animations. */
  stop(): this
  /** Stop animations for the given keys. */
  stop(keys: OneOrMore<string>): this
  /** Cancel all animations. */
  stop(cancel: boolean): this
  /** Cancel animations for the given keys. */
  stop(cancel: boolean, keys: OneOrMore<string>): this
  /** Stop some or all animations. */
  stop(keys?: OneOrMore<string>): this
  /** Cancel some or all animations. */
  stop(cancel: boolean, keys?: OneOrMore<string>): this
  /** @internal */
  stop(arg?: boolean | OneOrMore<string>, keys?: OneOrMore<string>) {
    if (arg !== !!arg) {
      keys = arg as OneOrMore<string>
    }
    if (keys) {
      const springs = this.springs as Lookup<SpringValue>
      each(toArray(keys), key => springs[key].stop(!!arg))
    } else {
      stopAsync(this._state, this._lastAsyncId)
      this.each(spring => spring.stop(!!arg))
    }
    return this
  }

  /** Freeze the active animation in time */
  pause(keys?: OneOrMore<string>) {
    if (is.und(keys)) {
      this.start({ pause: true })
    } else {
      const springs = this.springs as Lookup<SpringValue>
      each(toArray(keys), key => springs[key].pause())
    }
    return this
  }

  /** Resume the animation if paused. */
  resume(keys?: OneOrMore<string>) {
    if (is.und(keys)) {
      this.start({ pause: false })
    } else {
      const springs = this.springs as Lookup<SpringValue>
      each(toArray(keys), key => springs[key].resume())
    }
    return this
  }

  /** Call a function once per spring value */
  each(iterator: (spring: SpringValue, key: string) => void) {
    eachProp(this.springs, iterator as any)
  }

  /** @internal Called at the end of every animation frame */
  protected _onFrame() {
    const { onStart, onChange, onRest } = this._events

    const active = this._active.size > 0
    const changed = this._changed.size > 0

    if ((active && !this._started) || (changed && !this._started)) {
      this._started = true
      flush(onStart, ([onStart, result]) => {
        result.value = this.get()
        onStart(result, this, this._item)
      })
    }

    const idle = !active && this._started
    const values = changed || (idle && onRest.size) ? this.get() : null

    if (changed && onChange.size) {
      flush(onChange, ([onChange, result]) => {
        result.value = values
        onChange(result, this, this._item)
      })
    }

    // The "onRest" queue is only flushed when all springs are idle.
    if (idle) {
      this._started = false
      flush(onRest, ([onRest, result]) => {
        result.value = values
        onRest(result, this, this._item)
      })
    }
  }

  /** @internal */
  eventObserved(event: FrameValue.Event) {
    if (event.type == 'change') {
      this._changed.add(event.parent)
      if (!event.idle) {
        this._active.add(event.parent)
      }
    } else if (event.type == 'idle') {
      this._active.delete(event.parent)
    }
    // The `onFrame` handler runs when a parent is changed or idle.
    else return
    raf.onFrame(this._onFrame)
  }
}

/**
 * Warning: Props might be mutated.
 */
export function flushUpdateQueue(
  ctrl: Controller<any>,
  queue: ControllerQueue
) {
  return Promise.all(
    queue.map(props => flushUpdate(ctrl, props))
  ).then(results => getCombinedResult(ctrl, results))
}

/**
 * Warning: Props might be mutated.
 *
 * Process a single set of props using the given controller.
 *
 * The returned promise resolves to `true` once the update is
 * applied and any animations it starts are finished without being
 * stopped or cancelled.
 */
export async function flushUpdate(
  ctrl: Controller<any>,
  props: ControllerQueue[number],
  isLoop?: boolean
): AsyncResult {
  const { keys, to, from, loop, onRest, onResolve } = props
  const defaults = is.obj(props.default) && props.default

  // Looping must be handled in this function, or else the values
  // would end up looping out-of-sync in many common cases.
  if (loop) {
    props.loop = false
  }

  // Treat false like null, which gets ignored.
  if (to === false) props.to = null
  if (from === false) props.from = null

  const asyncTo = is.arr(to) || is.fun(to) ? to : undefined
  if (asyncTo) {
    props.to = undefined
    props.onRest = undefined
    if (defaults) {
      defaults.onRest = undefined
    }
  }
  // For certain events, use batching to prevent multiple calls per frame.
  // However, batching is avoided when the `to` prop is async, because any
  // event props are used as default props instead.
  else {
    each(BATCHED_EVENTS, key => {
      const handler: any = props[key]
      if (is.fun(handler)) {
        const queue = ctrl['_events'][key]
        props[key] = (({ finished, cancelled }: AnimationResult) => {
          const result = queue.get(handler)
          if (result) {
            if (!finished) result.finished = false
            if (cancelled) result.cancelled = true
          } else {
            // The "value" is set before the "handler" is called.
            queue.set(handler, {
              value: null,
              finished: finished || false,
              cancelled: cancelled || false,
            })
          }
        }) as any

        // Avoid using a batched `handler` as a default prop.
        if (defaults) {
          defaults[key] = props[key] as any
        }
      }
    })
  }

  const state = ctrl['_state']

  // Pause/resume the `asyncTo` when `props.pause` is true/false.
  if (props.pause === !state.paused) {
    state.paused = props.pause
    flushCalls(props.pause ? state.pauseQueue : state.resumeQueue)
  }
  // When a controller is paused, its values are also paused.
  else if (state.paused) {
    props.pause = true
  }

  const promises: AsyncResult[] = (keys || Object.keys(ctrl.springs)).map(key =>
    ctrl.springs[key]!.start(props as any)
  )

  const cancel =
    props.cancel === true || getDefaultProp(props, 'cancel') === true

  if (asyncTo || (cancel && state.asyncId)) {
    promises.push(
      scheduleProps(++ctrl['_lastAsyncId'], {
        props,
        state,
        actions: {
          pause: noop,
          resume: noop,
          start(props, resolve) {
            if (cancel) {
              stopAsync(state, ctrl['_lastAsyncId'])
              resolve(getCancelledResult(ctrl))
            } else {
              props.onRest = onRest
              resolve(runAsync(asyncTo!, props, state, ctrl))
            }
          },
        },
      })
    )
  }

  // Pause after updating each spring, so they can be resumed separately
  // and so their default `pause` and `cancel` props are updated.
  if (state.paused) {
    // Ensure `this` must be resumed before the returned promise
    // is resolved and before starting the next `loop` repetition.
    await new Promise<void>(resume => {
      state.resumeQueue.add(resume)
    })
  }

  const result = getCombinedResult<any>(ctrl, await Promise.all(promises))
  if (loop && result.finished && !(isLoop && result.noop)) {
    const nextProps = createLoopUpdate(props, loop, to)
    if (nextProps) {
      prepareKeys(ctrl, [nextProps])
      return flushUpdate(ctrl, nextProps, true)
    }
  }
  if (onResolve) {
    raf.batchedUpdates(() => onResolve(result, ctrl, ctrl.item))
  }
  return result
}

/**
 * From an array of updates, get the map of `SpringValue` objects
 * by their keys. Springs are created when any update wants to
 * animate a new key.
 *
 * Springs created by `getSprings` are neither cached nor observed
 * until they're given to `setSprings`.
 */
export function getSprings<State extends Lookup>(
  ctrl: Controller<State>,
  props?: OneOrMore<ControllerUpdate<State>>
) {
  const springs = { ...ctrl.springs }
  if (props) {
    each(toArray(props), (props: any) => {
      if (is.und(props.keys)) {
        props = createUpdate(props)
      }
      if (!is.obj(props.to)) {
        // Avoid passing array/function to each spring.
        props = { ...props, to: undefined }
      }
      prepareSprings(springs as any, props, key => {
        return createSpring(key)
      })
    })
  }
  return springs
}

/**
 * Tell a controller to manage the given `SpringValue` objects
 * whose key is not already in use.
 */
export function setSprings(
  ctrl: Controller,
  springs: SpringValues<UnknownProps>
) {
  eachProp(springs, (spring, key) => {
    if (!ctrl.springs[key]) {
      ctrl.springs[key] = spring
      addFluidObserver(spring, ctrl)
    }
  })
}

function createSpring(key: string, observer?: FluidObserver<FrameValue.Event>) {
  const spring = new SpringValue()
  spring.key = key
  if (observer) {
    addFluidObserver(spring, observer)
  }
  return spring
}

/**
 * Ensure spring objects exist for each defined key.
 *
 * Using the `props`, the `Animated` node of each `SpringValue` may
 * be created or updated.
 */
function prepareSprings(
  springs: SpringValues,
  props: ControllerQueue[number],
  create: (key: string) => SpringValue
) {
  if (props.keys) {
    each(props.keys, key => {
      const spring = springs[key] || (springs[key] = create(key))
      spring['_prepareNode'](props)
    })
  }
}

/**
 * Ensure spring objects exist for each defined key, and attach the
 * `ctrl` to them for observation.
 *
 * The queue is expected to contain `createUpdate` results.
 */
function prepareKeys(ctrl: Controller<any>, queue: ControllerQueue[number][]) {
  each(queue, props => {
    prepareSprings(ctrl.springs, props, key => {
      return createSpring(key, ctrl)
    })
  })
}
