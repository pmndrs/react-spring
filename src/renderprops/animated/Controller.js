import Animated from './Animated'
import AnimatedValue from './AnimatedValue'
import AnimatedArray from './AnimatedArray'
import { now, colorNames, forceImmediate } from './Globals'
import { addController, removeController } from './FrameLoop'
import {
  interpolateTo,
  withDefault,
  toArray,
  getValues,
  callProp,
  shallowEqual,
} from '../shared/helpers'

export default class Controller {
  constructor(
    props,
    config = { native: true, interpolateTo: true, autoStart: true }
  ) {
    this.dependents = new Set()
    this.isActive = false
    this.hasChanged = false
    this.props = {}
    this.merged = {}
    this.animations = {}
    this.interpolations = {}
    this.animatedProps = {}
    this.configs = []
    this.frame = undefined
    this.startTime = undefined
    this.lastTime = undefined
    this.update({ ...props, ...config })
  }

  update(props, ...start) {
    this.props = { ...this.props, ...props }
    let {
      from = {},
      to = {},
      config = {},
      delay = 0,
      reverse,
      attach,
      reset,
      immediate,
      autoStart,
      ref,
    } = this.props.interpolateTo ? interpolateTo(this.props) : this.props

    // Reverse values when requested
    if (reverse) {
      ;[from, to] = [to, from]
    }

    if (forceImmediate) {
      immediate = true
    }

    this.hasChanged = false
    // Attachment handling, trailed springs can "attach" themselves to a previous spring
    let target = attach && attach(this)

    // Reset merged props when necessary
    let extra = reset ? {} : this.merged
    // This will collect all props that were ever set
    this.merged = { ...from, ...extra, ...to }
    // Reduces input { name: value } pairs into animated values
    this.animations = Object.entries(this.merged).reduce(
      (acc, [name, value], i) => {
        // Issue cached entries, except on reset
        let entry = (!reset && acc[name]) || {}

        // Figure out what the value is supposed to be
        const isNumber = typeof value === 'number'
        const isString =
          typeof value === 'string' &&
          !value.startsWith('#') &&
          !/\d/.test(value) &&
          !colorNames[value]
        const isArray = !isNumber && !isString && Array.isArray(value)

        let fromValue = from[name] !== undefined ? from[name] : value
        let toValue = isNumber || isArray ? value : isString ? value : 1
        let toConfig = callProp(config, name)
        if (target) toValue = target.animations[name].parent

        // Detect changes, animated values will be checked in the raf-loop
        if (toConfig.decay !== void 0 || !shallowEqual(entry.changes, value)) {
          this.hasChanged = true
          let parent, interpolation
          if (isNumber || isString)
            parent = interpolation =
              entry.parent || new AnimatedValue(fromValue)
          else if (isArray)
            parent = interpolation =
              entry.parent || new AnimatedArray(fromValue)
          else {
            const prev =
              entry.interpolation &&
              entry.interpolation.calc(entry.parent.value)
            if (entry.parent) {
              parent = entry.parent
              parent.setValue(0, false)
            } else parent = new AnimatedValue(0)
            const range = {
              output: [prev !== void 0 ? prev : fromValue, value],
            }
            if (entry.interpolation) {
              interpolation = entry.interpolation
              entry.interpolation.updateConfig(range)
            } else interpolation = parent.interpolate(range)
          }

          // Set immediate values
          if (callProp(immediate, name)) parent.setValue(value, false)

          // Reset animated values
          const animatedValues = toArray(parent.getPayload())
          animatedValues.forEach(value => value.prepare(this))

          return {
            ...acc,
            [name]: {
              ...entry,
              name,
              parent,
              interpolation,
              animatedValues,
              changes: value,
              fromValues: toArray(parent.getValue()),
              toValues: toArray(target ? toValue.getPayload() : toValue),
              immediate: callProp(immediate, name),
              delay: withDefault(toConfig.delay, delay || 0),
              initialVelocity: withDefault(toConfig.velocity, 0),
              clamp: withDefault(toConfig.clamp, false),
              precision: withDefault(toConfig.precision, 0.01),
              tension: withDefault(toConfig.tension, 170),
              friction: withDefault(toConfig.friction, 26),
              mass: withDefault(toConfig.mass, 1),
              duration: toConfig.duration,
              easing: withDefault(toConfig.easing, t => t),
              decay: toConfig.decay,
            },
          }
        } else return acc
      },
      this.animations
    )

    if (this.hasChanged) {
      this.configs = getValues(this.animations)
      this.animatedProps = {}
      this.interpolations = {}
      for (let key in this.animations) {
        this.interpolations[key] = this.animations[key].interpolation
        this.animatedProps[key] = this.animations[key].interpolation.getValue()
      }
    }

    // TODO: clean up ref in controller
    if (!ref && (autoStart || start.length)) this.start(...start)
    const [onEnd, onUpdate] = start
    this.onEnd = typeof onEnd === 'function' && onEnd
    this.onUpdate = onUpdate
    return this.getValues()
  }

  start(onEnd, onUpdate) {
    this.startTime = now()
    if (this.isActive) this.stop()
    this.isActive = true
    this.onEnd = typeof onEnd === 'function' && onEnd
    this.onUpdate = onUpdate
    if (this.props.onStart) this.props.onStart()
    addController(this)
    return new Promise(res => (this.resolve = res))
  }

  stop(finished = false) {
    // Reset collected changes since the animation has been stopped cold turkey
    if (finished)
      getValues(this.animations).forEach(a => (a.changes = undefined))
    this.debouncedOnEnd({ finished })
  }

  destroy() {
    removeController(this)
    this.props = {}
    this.merged = {}
    this.animations = {}
    this.interpolations = {}
    this.animatedProps = {}
    this.configs = []
  }

  debouncedOnEnd(result) {
    removeController(this)
    this.isActive = false
    const onEnd = this.onEnd
    this.onEnd = null
    if (onEnd) onEnd(result)
    if (this.resolve) this.resolve()
    this.resolve = null
  }

  getValues = () =>
    this.props.native ? this.interpolations : this.animatedProps
}
