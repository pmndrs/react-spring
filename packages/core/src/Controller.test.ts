import { Controller } from './Controller'

const frameLength = 1000 / 60

describe('Controller', () => {
  it('can animate a number', async () => {
    const ctrl = new Controller({ x: 0 })
    ctrl.start({ x: 100 })

    await advanceUntilIdle()
    const frames = getFrames(ctrl)
    expect(frames).toMatchSnapshot()

    // The first frame should *not* be the from value.
    expect(frames[0]).not.toEqual({ x: 0 })

    // The last frame should be the goal value.
    expect(frames.slice(-1)[0]).toEqual({ x: 100 })
  })

  it('can animate an array of numbers', async () => {
    const config = { precision: 0.005 }
    const ctrl = new Controller<{ x: [number, number] }>({ x: [1, 2], config })
    ctrl.start({ x: [5, 10] })

    await advanceUntilIdle()
    const frames = getFrames(ctrl)
    expect(frames).toMatchSnapshot()

    // The last frame should be the goal value.
    expect(frames.slice(-1)[0]).toEqual({ x: [5, 10] })

    // The 2nd value is always ~2x the 1st value (within the defined precision).
    const factors = frames.map(frame => frame.x[1] / frame.x[0])
    expect(
      factors.every(factor => Math.abs(2 - factor) < config.precision)
    ).toBeTruthy()
  })

  describe('when the "to" prop is an async function', () => {
    it('acts strangely without the "from" prop', async () => {
      const ctrl = new Controller<{ x: number }>()

      const { springs } = ctrl
      ctrl.start({
        to: async update => {
          // The spring does not exist yet!
          expect(springs.x).toBeUndefined()

          // Any values passed here are treated as "from" values,
          // because no "from" prop was ever given.
          const p1 = update({ x: 1 })
          // Now the spring exists!
          expect(springs.x).toBeDefined()
          // But the spring is idle!
          expect(springs.x.idle).toBeTruthy()

          // This call *will* start an animation!
          const p2 = update({ x: 2 })
          expect(springs.x.idle).toBeFalsy()

          await Promise.all([p1, p2])
        },
      })

      await advanceUntilIdle()

      // Since we call `update` twice, frames are generated!
      expect(getFrames(ctrl)).toMatchSnapshot()
    })

    it('can be cancelled', async () => {
      const ctrl = new Controller({ from: { x: 0 } })
      const promise = ctrl.start({
        to: async next => {
          while (true) {
            await next({ x: 1, reset: true })
          }
        },
      })
      const { x } = ctrl.springs
      await advanceUntilValue(x, 0.5)
      await ctrl.start({
        cancel: true,
      })
      expect(ctrl.idle).toBeTruthy()
      expect((await promise).cancelled).toBeTruthy()
    })
  })

  describe('when the "onStart" prop is defined', () => {
    it('is called once per "start" call maximum', async () => {
      const ctrl = new Controller({ x: 0, y: 0 })

      const onStart = jest.fn()
      ctrl.start({
        x: 1,
        y: 1,
        onStart,
      })

      await advanceUntilIdle()
      expect(onStart).toBeCalledTimes(1)
    })

    it('can be different per key', async () => {
      const ctrl = new Controller({ x: 0, y: 0 })

      const onStart1 = jest.fn()
      ctrl.start({ x: 1, onStart: onStart1 })

      const onStart2 = jest.fn()
      ctrl.start({ y: 1, onStart: onStart2 })

      await advanceUntilIdle()
      expect(onStart1).toBeCalledTimes(1)
      expect(onStart2).toBeCalledTimes(1)
    })
  })

  describe('the "loop" prop', () => {
    it('can be combined with the "reverse" prop', async () => {
      const ctrl = new Controller({
        t: 1,
        from: { t: 0 },
        config: { duration: frameLength * 3 },
      })

      const { t } = ctrl.springs
      expect(t.get()).toBe(0)

      await advanceUntilIdle()
      expect(t.get()).toBe(1)

      ctrl.start({
        loop: { reverse: true },
      })

      await advanceUntilValue(t, 0)
      await advanceUntilValue(t, 1)
      expect(getFrames(t)).toMatchSnapshot()
    })

    describe('used with multiple values', () => {
      it('loops all values at the same time', async () => {
        const ctrl = new Controller()

        ctrl.start({
          to: { x: 1, y: 1 },
          from: { x: 0, y: 0 },
          config: key => ({ frequency: key == 'x' ? 0.3 : 1 }),
          loop: true,
        })

        const { x, y } = ctrl.springs
        for (let i = 0; i < 2; i++) {
          await advanceUntilValue(y, 1)

          // Both values should equal their "from" value at the same time.
          expect(x.get()).toBe(x.animation.from)
          expect(y.get()).toBe(y.animation.from)
        }
      })
    })

    describe('used when "to" is', () => {
      describe('an async function', () => {
        it('calls the "to" function repeatedly', async () => {
          const ctrl = new Controller({ t: 0 })
          const { t } = ctrl.springs

          let loop = true
          let times = 2

          // Note: This example is silly, since you could use a for-loop
          // to more easily achieve the same result, but it tests the ability
          // to halt a looping script via the "loop" function prop.
          ctrl.start({
            loop: () => loop,
            to: async next => {
              await next({ t: 1 })
              await next({ t: 0 })

              if (times--) return
              loop = false
            },
          })

          await advanceUntilValue(t, 1)
          expect(t.idle).toBeFalsy()

          for (let i = 0; i < 2; i++) {
            await advanceUntilValue(t, 0)
            expect(t.idle).toBeFalsy()

            await advanceUntilValue(t, 1)
            expect(t.idle).toBeFalsy()
          }

          await advanceUntilValue(t, 0)
          expect(t.idle).toBeTruthy()
        })
      })

      describe('an array', () => {
        it('repeats the chain of updates', async () => {
          const ctrl = new Controller({ t: 0 })
          const { t } = ctrl.springs

          let loop = true
          const promise = ctrl.start({
            loop: () => {
              return loop
            },
            from: { t: 0 },
            to: [{ t: 1 }, { t: 2 }],
            config: { duration: 3000 / 60 },
          })

          for (let i = 0; i < 3; i++) {
            await advanceUntilValue(t, 2)
            expect(t.idle).toBeFalsy()

            // Run the first frame of the next loop.
            mockRaf.step()
          }

          loop = false

          await advanceUntilValue(t, 2)
          expect(t.idle).toBeTruthy()

          expect(await promise).toMatchObject({
            value: { t: 2 },
            finished: true,
          })
        })
      })
    })

    describe('used on a noop update', () => {
      it('does not loop', async () => {
        const ctrl = new Controller({ t: 0 })

        const loop = jest.fn(() => true)
        ctrl.start({ t: 0, loop })

        await advanceUntilIdle()
        expect(loop).toBeCalledTimes(0)
      })
    })
  })
})
