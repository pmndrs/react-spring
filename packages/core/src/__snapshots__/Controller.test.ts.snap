// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Controller can animate a number 1`] = `
[
  {
    "x": 2.263484330785799,
  },
  {
    "x": 7.615528894993947,
  },
  {
    "x": 14.72729545687759,
  },
  {
    "x": 22.67933293781441,
  },
  {
    "x": 30.85053191603684,
  },
  {
    "x": 38.83524421610096,
  },
  {
    "x": 46.38171205392548,
  },
  {
    "x": 53.346574480369156,
  },
  {
    "x": 59.66146523130329,
  },
  {
    "x": 65.30867180317823,
  },
  {
    "x": 70.30355736424129,
  },
  {
    "x": 74.68200657416291,
  },
  {
    "x": 78.49158337229619,
  },
  {
    "x": 81.78541406742545,
  },
  {
    "x": 84.61805634019797,
  },
  {
    "x": 87.04280232770594,
  },
  {
    "x": 89.11000586164683,
  },
  {
    "x": 90.8661309883985,
  },
  {
    "x": 92.35329941206051,
  },
  {
    "x": 93.60917483504707,
  },
  {
    "x": 94.66706719880813,
  },
  {
    "x": 95.55617327559861,
  },
  {
    "x": 96.30189477450246,
  },
  {
    "x": 96.92619326731115,
  },
  {
    "x": 97.4479544590839,
  },
  {
    "x": 97.88334387327176,
  },
  {
    "x": 98.2461428371077,
  },
  {
    "x": 98.54805845242963,
  },
  {
    "x": 98.7990045563195,
  },
  {
    "x": 99.0073529166361,
  },
  {
    "x": 99.18015536949562,
  },
  {
    "x": 99.3233385117055,
  },
  {
    "x": 99.4418730756026,
  },
  {
    "x": 99.53992035746593,
  },
  {
    "x": 99.62095813165256,
  },
  {
    "x": 99.68788842438258,
  },
  {
    "x": 99.74312938902625,
  },
  {
    "x": 99.78869335076726,
  },
  {
    "x": 99.8262528947063,
  },
  {
    "x": 99.85719667273446,
  },
  {
    "x": 99.8826764105673,
  },
  {
    "x": 100,
  },
]
`;

exports[`Controller can animate an array of numbers 1`] = `
[
  {
    "x": [
      1.0905393732314317,
      2.1810787464628634,
    ],
  },
  {
    "x": [
      1.3046211557997573,
      2.6092423115995147,
    ],
  },
  {
    "x": [
      1.589091818275103,
      3.178183636550206,
    ],
  },
  {
    "x": [
      1.9071733175125758,
      3.8143466350251516,
    ],
  },
  {
    "x": [
      2.2340212766414735,
      4.468042553282947,
    ],
  },
  {
    "x": [
      2.5534097686440376,
      5.106819537288075,
    ],
  },
  {
    "x": [
      2.8552684821570185,
      5.710536964314037,
    ],
  },
  {
    "x": [
      3.1338629792147654,
      6.267725958429531,
    ],
  },
  {
    "x": [
      3.3864586092521307,
      6.772917218504261,
    ],
  },
  {
    "x": [
      3.612346872127128,
      7.224693744254256,
    ],
  },
  {
    "x": [
      3.8121422945696497,
      7.624284589139299,
    ],
  },
  {
    "x": [
      3.987280262966515,
      7.97456052593303,
    ],
  },
  {
    "x": [
      4.139663334891849,
      8.279326669783698,
    ],
  },
  {
    "x": [
      4.271416562697019,
      8.542833125394038,
    ],
  },
  {
    "x": [
      4.3847222536079205,
      8.769444507215841,
    ],
  },
  {
    "x": [
      4.481712093108238,
      8.963424186216477,
    ],
  },
  {
    "x": [
      4.5644002344658725,
      9.128800468931745,
    ],
  },
  {
    "x": [
      4.634645239535937,
      9.269290479071874,
    ],
  },
  {
    "x": [
      4.694131976482419,
      9.388263952964838,
    ],
  },
  {
    "x": [
      4.7443669934018855,
      9.488733986803771,
    ],
  },
  {
    "x": [
      4.786682687952328,
      9.573365375904656,
    ],
  },
  {
    "x": [
      4.822246931023949,
      9.644493862047899,
    ],
  },
  {
    "x": [
      4.852075790980104,
      9.704151581960208,
    ],
  },
  {
    "x": [
      4.877047730692452,
      9.754095461384903,
    ],
  },
  {
    "x": [
      4.89791817836336,
      9.79583635672672,
    ],
  },
  {
    "x": [
      4.915333754930874,
      9.830667509861748,
    ],
  },
  {
    "x": [
      4.929845713484313,
      9.859691426968626,
    ],
  },
  {
    "x": [
      4.941922338097189,
      9.883844676194379,
    ],
  },
  {
    "x": [
      4.951960182252786,
      9.903920364505572,
    ],
  },
  {
    "x": [
      4.96029411666545,
      9.9205882333309,
    ],
  },
  {
    "x": [
      4.967206214779832,
      9.934412429559664,
    ],
  },
  {
    "x": [
      4.972933540468227,
      9.945867080936454,
    ],
  },
  {
    "x": [
      4.97767492302411,
      9.95534984604822,
    ],
  },
  {
    "x": [
      4.981596814298641,
      9.963193628597281,
    ],
  },
  {
    "x": [
      4.984838325266105,
      9.96967665053221,
    ],
  },
  {
    "x": [
      4.9875155369753035,
      9.975031073950607,
    ],
  },
  {
    "x": [
      4.989725175561048,
      9.979450351122097,
    ],
  },
  {
    "x": [
      4.991547734030691,
      9.983095468061382,
    ],
  },
  {
    "x": [
      4.993050115788252,
      9.986100231576504,
    ],
  },
  {
    "x": [
      4.994287866909378,
      9.988575733818756,
    ],
  },
  {
    "x": [
      4.995027593476066,
      9.990614112845385,
    ],
  },
  {
    "x": [
      4.995027593476066,
      9.992291713056268,
    ],
  },
  {
    "x": [
      4.995027593476066,
      9.99367173585381,
    ],
  },
  {
    "x": [
      4.995027593476066,
      9.994806458630702,
    ],
  },
  {
    "x": [
      5,
      10,
    ],
  },
]
`;

exports[`Controller the "loop" prop can be combined with the "reverse" prop 1`] = `
[
  0.33334,
  0.66668,
  1,
  0.66666,
  0.33331999999999995,
  0,
  0.33334,
  0.66668,
  1,
]
`;

exports[`Controller when the "to" prop is an async function acts strangely without the "from" prop 1`] = `
[
  {
    "x": 1.0226348433078583,
  },
  {
    "x": 1.0761552889499395,
  },
  {
    "x": 1.1472729545687759,
  },
  {
    "x": 1.226793329378144,
  },
  {
    "x": 1.3085053191603682,
  },
  {
    "x": 1.3883524421610094,
  },
  {
    "x": 1.4638171205392543,
  },
  {
    "x": 1.5334657448036912,
  },
  {
    "x": 1.5966146523130327,
  },
  {
    "x": 1.6530867180317823,
  },
  {
    "x": 1.7030355736424132,
  },
  {
    "x": 1.74682006574163,
  },
  {
    "x": 1.7849158337229636,
  },
  {
    "x": 1.817854140674256,
  },
  {
    "x": 1.8461805634019814,
  },
  {
    "x": 1.870428023277061,
  },
  {
    "x": 1.8911000586164695,
  },
  {
    "x": 1.9086613098839855,
  },
  {
    "x": 1.9235329941206056,
  },
  {
    "x": 1.9360917483504718,
  },
  {
    "x": 1.9466706719880824,
  },
  {
    "x": 1.9555617327559878,
  },
  {
    "x": 1.9630189477450264,
  },
  {
    "x": 1.9692619326731133,
  },
  {
    "x": 1.9744795445908405,
  },
  {
    "x": 1.9788334387327189,
  },
  {
    "x": 1.9824614283710786,
  },
  {
    "x": 1.9854805845242978,
  },
  {
    "x": 1.9879900455631967,
  },
  {
    "x": 1.9900735291663625,
  },
  {
    "x": 1.991801553694958,
  },
  {
    "x": 1.9932333851170567,
  },
  {
    "x": 1.9944187307560275,
  },
  {
    "x": 1.9953992035746602,
  },
  {
    "x": 1.9962095813165261,
  },
  {
    "x": 1.9968788842438259,
  },
  {
    "x": 1.997431293890262,
  },
  {
    "x": 1.9978869335076728,
  },
  {
    "x": 1.998262528947063,
  },
  {
    "x": 1.9985719667273445,
  },
  {
    "x": 1.9988267641056732,
  },
  {
    "x": 2,
  },
]
`;
